### PR TITLE
Add debug mode and integrate with existing logging

### PR DESCRIPTION
**Pull request by adamkpickering**
For #538. This takes care of only part of it.

### CODE REVIEW

**Comment by ericpromislow** [comment]
There's no reason to restart everything when the debug state changes, so I would make a change like this (which could be used for other settings as well):
```
diff --git a/background.ts b/background.ts
index 21afd27f..03886b1f 100644
--- a/background.ts
+++ b/background.ts
@@ -322,7 +322,9 @@ function writeSettings(arg: RecursivePartial<settings.Settings>) {
   _.merge(cfg, arg);
   settings.save(cfg);
   mainEvents.emit('settings-update', cfg);
-  Electron.ipcMain.emit('k8s-restart-required');
+  if (_.difference(Object.keys(arg), ['debug']).length > 0) {
+    Electron.ipcMain.emit('k8s-restart-required');
+  }
 }
 
 Electron.ipcMain.handle('settings-write', (event, arg) => {
```

Also, I'm currently alone in this, but I prefer putting the issue no at the start of the branch name. `add-debug-mode` looks better (to me) when it's `538-add-debug-mode`.

**Review by ericpromislow — CHANGES_REQUESTED** [review]
Just the request to not restart when the debug setting changes.

File: src/utils/logging.ts
Comment (mook-as) [inline src/utils/logging.ts:121]: Given that _most_ importers only need the main `Logging`, wouldn't `export default` make sense here? `background.ts` can do `import Logging, { setLogLevel } from …`

See `src/window/index.ts` for an example: `import Electron, { BrowserWindow, app, shell } from 'electron';`

File: background.ts
Comment (mook-as) [inline background.ts:299]: Is this debug logging required?

File: src/config/settings.ts
Comment (mook-as) [inline src/config/settings.ts:21]: Not sure we need the version bump (that is, I don't think we need an explicit migration step — didn't add one to `updateTable` either).

File: src/components/DebugMode.vue
Comment (mook-as) [inline src/components/DebugMode.vue:3]: What does `<div class="checkbox">` do? I couldn't find any rules that use it (and it's unclear why it needs to be a child item).

I also couldn't spot what `<div class="system-preferences">` did.

But then, we end up with just a `<Checkbox>` by itself, with no useful extra logic, so we can get rid of this whole component…

File: src/pages/Troubleshooting.vue
Comment (mook-as) [inline src/pages/Troubleshooting.vue:117]: I can't decide if restarting here is a good thing — if the cluster is working fine, do we need to bring it down? But I guess if the user is flipping switches it's _probably_ not working fine?

**Review by mook-as — CHANGES_REQUESTED** [review]

File: src/utils/logging.ts
Comment (adamkpickering) [reply src/utils/logging.ts:121]: Ah I didn't know you could do that. I'll implement this.

File: background.ts
Comment (adamkpickering) [reply background.ts:299]: I mean, no logging is required. But given that this epic is about making debugging information available, this is the sort of thing that we would want to add, unless I'm mistaken?

File: src/config/settings.ts
Comment (adamkpickering) [reply src/config/settings.ts:21]: :+1: implemented

Comment (ericpromislow) [reply src/config/settings.ts:21]: Normally, we only need version bumps for settings when we introduce a breaking change in the settings structure. Don't see any in this PR.

File: src/components/DebugMode.vue
Comment (adamkpickering) [reply src/components/DebugMode.vue:3]: I wondered if it was possible to just use `<Checkbox>`, and then I didn't try it out. I tried it now and it looks good. Good thing you mentioned it :)

File: background.ts
Comment (ericpromislow) [reply background.ts:299]: I'd say debug-logging a `settings-write` is more useful than a `settings-read`, and we can leave it in the code.

File: src/pages/Troubleshooting.vue
Comment (mook-as) [inline src/pages/Troubleshooting.vue:112]: Extraneous logging (that wouldn't be useful to debug customer issues).

**Review by mook-as — DISMISSED** [review]

File: src/config/settings.ts
Comment (ericpromislow) [inline src/config/settings.ts:52]: This is almost a coin flip, but now that we have this, I think it should be true by default, because it's quite common that someone will run into problems early on, and it would be useful to have the debug statements present.

I'm seeing a work item to add debug statements at many important points in the code, such as starting up K8s on lima or WSL.

File: src/pages/Troubleshooting.vue
Comment (ericpromislow) [inline src/pages/Troubleshooting.vue:70]: Please talk to Everton on how to do this asynchronously. He's got a PR on this that's waiting on git actions before it gets merged

**Review by ericpromislow — CHANGES_REQUESTED** [review]
Two suggestions:
1. Move the debug stmt to where settings get written -- easy to test, and happens less often than reading.
2. See PR 949 and/or talk to Everton about the async message handling. In fact, I think his PR will break this code when it merges because only async `settings-read` will be supported.

File: src/config/settings.ts
Comment (mook-as) [reply src/config/settings.ts:52]: I'd vote the other way — if debug logging is on by default, we might as well not have it.  It should be an option to get _additional_ output for troubleshooting, and never set by default. (See also the terrible mess that is Cloud Foundry logging.)

Comment (ericpromislow) [reply src/config/settings.ts:52]: Good point, although I admit I was thinking exactly of that CF logging. They even have unit tests that verify certain logs are being issued.

The issue with having the debug setting off initially is say the user on a first or early use runs into a weird problem and contacts us immediately. We can ask for a log file that might have something helpful in it. If it's a flake, and doesn't happen on a subsequent run after they turn debugging on, we've missed potentially useful info.  But even while writing this I was thinking it's such a long shot. So sure, leave the default at false.

**Review by mook-as — CHANGES_REQUESTED** [review]
Just need to fix the `sendSync` in `Troubleshooting.vue`.

**Comment by adamkpickering** [comment]
> Just need to fix the `sendSync` in `Troubleshooting.vue`.

If I do that, I'll need to make a change in `background.ts`, `Images.vue`, `K8s.vue`, and `PortForwarding.vue` as well. I figured that was better kept for another PR.

**Comment by mook-as** [comment]
No, just fix the one you changed _here_… as in, just don't add more uses of `sendSync`; we don't need to worry about existing ones yet.

**Review by mook-as — APPROVED** [review]
Sorry about that — I misread the diff; the `sendSync` was existing.

**Review by ericpromislow — APPROVED** [review]
LGTM

**Pull request merged**
